core/runtime/exception_state.rs

```diff
@@ -0,0 +1,127 @@
+use crate::error::exception_to_err_result;
+use anyhow::Error;
+use std::cell::Cell;
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::rc::Rc;
+
+#[derive(Default)]
+pub(crate) struct ExceptionState {
+  // TODO(nayeemrmn): This is polled in `exception_to_err_result()` which is
+  // flimsy. Try to poll it similarly to `pending_promise_rejections`.
+  dispatched_exception: Cell<Option<v8::Global<v8::Value>>>,
+  dispatched_exception_is_promise: Cell<bool>,
+  pub(crate) pending_promise_rejections:
+    RefCell<VecDeque<(v8::Global<v8::Promise>, v8::Global<v8::Value>)>>,
+  pub(crate) js_build_custom_error_cb:
+    RefCell<Option<Rc<v8::Global<v8::Function>>>>,
+  pub(crate) js_promise_reject_cb:
+    RefCell<Option<Rc<v8::Global<v8::Function>>>>,
+  pub(crate) js_format_exception_cb:
+    RefCell<Option<Rc<v8::Global<v8::Function>>>>,
+}
+
+impl ExceptionState {
+  pub fn destroy(&self) {
+    self.js_build_custom_error_cb.borrow_mut().take();
+    self.js_promise_reject_cb.borrow_mut().take();
+    self.js_format_exception_cb.borrow_mut().take();
+    self.pending_promise_rejections.borrow_mut().clear();
+    self.dispatched_exception.set(None);
+  }
+
+  pub(crate) fn clear_error(&self) {
+    self.dispatched_exception_is_promise.set(false);
+    self.dispatched_exception.set(None);
+  }
+
+  pub(crate) fn has_dispatched_exception(&self) -> bool {
+    // SAFETY: we limit access to this cell to this method only
+    unsafe {
+      self
+        .dispatched_exception
+        .as_ptr()
+        .as_ref()
+        .unwrap_unchecked()
+        .is_some()
+    }
+  }
+
+  pub(crate) fn set_dispatched_exception(
+    &self,
+    exception: v8::Global<v8::Value>,
+    promise: bool,
+  ) {
+    self.dispatched_exception.set(Some(exception));
+    self.dispatched_exception_is_promise.set(promise);
+  }
+
+  pub(crate) fn check_exception_condition(
+    &self,
+    scope: &mut v8::HandleScope,
+  ) -> Result<(), Error> {
+    // TODO(mmastrac): rewrite before landing
+    if let Some(exception) = self.get_dispatched_exception_as_local(scope) {
+      exception_to_err_result(
+        scope,
+        exception,
+        self.dispatched_exception_is_promise.get(),
+        true,
+      )
+    } else {
+      Ok(())
+    }
+  }
+
+  pub(crate) fn is_dispatched_exception_promise(&self) -> bool {
+    self.dispatched_exception_is_promise.get()
+  }
+
+  pub(crate) fn get_dispatched_exception_as_local<'s>(
+    &self,
+    scope: &mut v8::HandleScope<'s>,
+  ) -> Option<v8::Local<'s, v8::Value>> {
+    // SAFETY: we limit access to this cell to this method only
+    unsafe {
+      self
+        .dispatched_exception
+        .as_ptr()
+        .as_ref()
+        .unwrap_unchecked()
+    }
+    .as_ref()
+    .map(|global| v8::Local::new(scope, global))
+  }
+
+  /// Calls the project rejection callback, if one exists.
```

Comment:
What's "project"?

Reply:
typo -- should be promise.

Reply:
Fixed, and renamed this method to `track_promise_rejection` since it's not actually calling a callback any more as well.